### PR TITLE
[854E] Update dependencies 04/24

### PR DESCRIPTION
**Pull request by jvega190**
https://github.com/craftersoftware/craftercms/issues/854

### CODE REVIEW

File: ui/app/src/components/GlobalApp/GlobalApp.tsx [hunk 1]
```diff
@@ -52,11 +60,79 @@ interface GlobalAppProps {
...
+
+  const router = createHashRouter(
+    createRoutesFromElements(
+      <Route path="/" element={<GlobalApp {...props} />}>
+        <Route path="/sites" element={<SiteManagement />} />
+        <Route
+          path="/users"
+          element={<UserManagement passwordRequirementsMinComplexity={passwordRequirementsMinComplexity} />}
+        />
+        <Route path="/groups" element={<GroupManagement />} />
```
Comment (rart) [inline ui/app/src/components/GlobalApp/GlobalApp.tsx]: Is there a way of defining the routes without the instantiation of the component at the definition?

File: ui/app/src/components/GlobalApp/GlobalApp.tsx [hunk 2]
```diff
@@ -52,11 +60,79 @@ interface GlobalAppProps {
...
 
+export function GlobalAppRouterProvider(props: GlobalAppProps) {
+  const { passwordRequirementsMinComplexity } = props;
+  const globalNavigation = useGlobalNavigation();
+  const { classes } = useStyles();
+
+  const router = createHashRouter(
+    createRoutesFromElements(
+      <Route path="/" element={<GlobalApp {...props} />}>
+        <Route path="/sites" element={<SiteManagement />} />
```
Comment (rart) [inline ui/app/src/components/GlobalApp/GlobalApp.tsx:78]: Leave this route for _backward compat_, but let's modify the main route to be "/projects"

File: ui/app/src/components/GlobalApp/GlobalApp.tsx [hunk 3]
```diff
@@ -52,11 +60,79 @@ interface GlobalAppProps {
...
 }
 
+export function GlobalAppRouterProvider(props: GlobalAppProps) {
+  const { passwordRequirementsMinComplexity } = props;
+  const globalNavigation = useGlobalNavigation();
+  const { classes } = useStyles();
+
+  const router = createHashRouter(
+    createRoutesFromElements(
+      <Route path="/" element={<GlobalApp {...props} />}>
```
Comment (rart) [inline ui/app/src/components/GlobalApp/GlobalApp.tsx]: Let's create some sort of enum with the routes so we can use them across the various places. For example here and when using `navigate('/global-config') => navigate(Routes.GlobalConfig)`

File: ui/app/src/components/GlobalConfigManagement/GlobalConfigManagement.tsx [hunk 4]
```diff
@@ -86,14 +76,52 @@ export function GlobalConfigManagement() {
...
+          body: <FormattedMessage defaultMessage="You left unsaved changes. Go back and continue editing?" />,
+          onCancel: batchActions([closeConfirmDialog(), dispatchDOMEvent({ id: eventId, button: 'cancel' })]),
+          onOk: batchActions([closeConfirmDialog(), dispatchDOMEvent({ id: eventId, button: 'ok' })]),
+          okButtonText: <FormattedMessage defaultMessage="Continue editing" />,
+          cancelButtonText: <FormattedMessage defaultMessage="Discard changes" />
+        })
+      );
+      createCustomDocumentEventListener<{ button: 'ok' | 'cancel' }>(eventId, ({ button }) => {
+        if (button === 'ok') {
+          navigate('/global-config');
```
Comment (rart) [inline ui/app/src/components/GlobalConfigManagement/GlobalConfigManagement.tsx]: Use the enum.

Look at other places we might be using routes like Launcher and its subcomponents, etc

File: ui/app/src/components/GlobalConfigManagement/GlobalConfigManagement.tsx [hunk 5]
```diff
@@ -86,14 +76,52 @@ export function GlobalConfigManagement() {
...
+        }
+      });
+    }
+  });
+
+  useBeforeUnload((event) => {
+    if (hasChangesRef.current) {
+      event.preventDefault();
+    }
+  });
```
Comment (rart) [inline ui/app/src/components/GlobalConfigManagement/GlobalConfigManagement.tsx:124]: Can you specify any kind of message to display? How does this work?

File: ui/app/src/components/Preview/Preview.tsx [hunk 6]
```diff
@@ -22,9 +22,12 @@ import PreviewConcierge from '../PreviewConcierge/PreviewConcierge';
 import usePreviewUrlControl from '../../hooks/usePreviewUrlControl';
 import ICEToolsPanel from '../ICEToolsPanel';
 import Box from '@mui/material/Box';
+import { useLocation, useNavigate } from 'react-router-dom';
 
 function Preview(props) {
-  usePreviewUrlControl(props.history);
+  const location = useLocation();
+  const push = useNavigate();
+  usePreviewUrlControl({ location, push });
```
Comment (rart) [inline ui/app/src/components/Preview/Preview.tsx]: Might as well call these directly inside usePreviewUrlControl? Update changelog

File: ui/app/src/components/Search/utils.ts [hunk 7]
```diff
@@ -63,7 +62,6 @@ export const actionsToBeShown: AllItemActions[] = [
 ];
 
 export interface URLDrivenSearchProps {
-  history: History;
```
Comment (rart) [inline ui/app/src/components/Search/utils.ts:66]: Update changelog

File: ui/app/src/components/SiteConfigurationManagement/SiteConfigurationManagement.tsx [hunk 8]
```diff
@@ -74,23 +73,33 @@ import { MAX_CONFIG_SIZE, UNDEFINED } from '../../utils/constants';
...
   onSubmittingAndOrPendingChange?(value: onSubmittingAndOrPendingChangeProps): void;
 }
 
 export function SiteConfigurationManagement(props: SiteConfigurationManagementProps) {
-  const { embedded, showAppsButton, onSubmittingAndOrPendingChange, isSubmitting } = props;
+  const { embedded, showAppsButton, onSubmittingAndOrPendingChange, isSubmitting, mountMode } = props;
   const site = useActiveSiteId();
+  const { username } = useActiveUser();
+  const sessionStorageKey = `craftercms.${username}.projectToolsConfigurationData.${site}`;
+  const sessionData = JSON.parse(sessionStorage.getItem(sessionStorageKey) ?? '{}');
```
Comment (rart) [inline ui/app/src/components/SiteConfigurationManagement/SiteConfigurationManagement.tsx]: x

File: ui/app/src/components/SiteConfigurationManagement/SiteConfigurationManagement.tsx [hunk 9]
```diff
@@ -74,23 +73,33 @@ import { MAX_CONFIG_SIZE, UNDEFINED } from '../../utils/constants';
...
   const baseUrl = useSelection<string>((state) => state.env.authoringBase);
   const { classes, cx: clsx } = useStyles();
   const { formatMessage } = useIntl();
   const [environment, setEnvironment] = useState<string>();
   const [files, setFiles] = useState<SiteConfigurationFileWithId[]>();
-  const [selectedConfigFile, setSelectedConfigFile] = useState<SiteConfigurationFileWithId>(null);
+  const [selectedConfigFile, setSelectedConfigFile] = useState<SiteConfigurationFileWithId>(
+    sessionData.selectedConfigFile ?? null
+  );
```
Comment (rart) [inline ui/app/src/components/SiteConfigurationManagement/SiteConfigurationManagement.tsx]:
```suggestion
  const [selectedConfigFile, setSelectedConfigFile] = useState<SiteConfigurationFileWithId>(() => 
    JSON.parse(sessionStorage.getItem(sessionStorageKey))?.selectedConfigFile ?? null;
  );
```

File: ui/app/src/components/SiteConfigurationManagement/SiteConfigurationManagement.tsx [hunk 10]
```diff
@@ -106,44 +115,20 @@ export function SiteConfigurationManagement(props: SiteConfigurationManagementPr
   const [disabledSaveButton, setDisabledSaveButton] = useState(true);
   const [confirmDialogProps, setConfirmDialogProps] = useState<ConfirmDialogProps>(null);
   const [keyword, setKeyword] = useState('');
-  const history = useHistory();
   const dispatch = useDispatch();
-  const disableBlocking = useRef(false);
+  const refs = useUpdateRefs({
+    disabledSaveButton,
+    selectedConfigFile,
+    sessionData
```
Comment (rart) [inline ui/app/src/components/SiteConfigurationManagement/SiteConfigurationManagement.tsx]: x

File: ui/app/src/components/SiteConfigurationManagement/SiteConfigurationManagement.tsx [hunk 11]
```diff
@@ -179,8 +193,15 @@ export function SiteConfigurationManagement(props: SiteConfigurationManagementPr
         ignoreEnv ? null : environment
       ).subscribe({
         next(xml) {
-          setSelectedConfigFileXml(xml ?? '');
-          setContentSize(xml?.length ?? 0);
+          if (refs.current.sessionData.content) {
```
Comment (rart) [inline ui/app/src/components/SiteConfigurationManagement/SiteConfigurationManagement.tsx]:
```suggestion
          const sessionData = get(sessionStorageKey)
          if (sessionData.content) {
```

File: ui/app/src/components/SiteConfigurationManagement/SiteConfigurationManagement.tsx [hunk 12]
```diff
@@ -156,6 +141,35 @@ export function SiteConfigurationManagement(props: SiteConfigurationManagementPr
...
+          body: <FormattedMessage defaultMessage="You left unsaved changes. Go back and continue editing?" />,
+          onCancel: batchActions([closeConfirmDialog(), dispatchDOMEvent({ id: eventId, button: 'cancel' })]),
+          onOk: batchActions([closeConfirmDialog(), dispatchDOMEvent({ id: eventId, button: 'ok' })]),
+          okButtonText: <FormattedMessage defaultMessage="Continue editing" />,
+          cancelButtonText: <FormattedMessage defaultMessage="Discard changes" />
+        })
+      );
+      createCustomDocumentEventListener<{ button: 'ok' | 'cancel' }>(eventId, ({ button }) => {
+        if (button === 'ok') {
+          navigate('/configuration');
```
Comment (rart) [inline ui/app/src/components/SiteConfigurationManagement/SiteConfigurationManagement.tsx]:
```suggestion
          navigate(Routes.Configuration);
```

File: ui/app/src/components/GlobalApp/GlobalApp.tsx [hunk 13]
```diff
@@ -52,11 +60,79 @@ interface GlobalAppProps {
   footerHtml: string;
 }
 
+export function GlobalAppRouterProvider(props: GlobalAppProps) {
```
Comment (rart) [inline ui/app/src/components/GlobalApp/GlobalApp.tsx]:
```suggestion
export function GlobalApp(props: GlobalAppProps) {
```

File: ui/app/src/components/GlobalApp/GlobalApp.tsx [hunk 14]
```diff
@@ -52,11 +60,79 @@ interface GlobalAppProps {
...
+          }
+        />
+      </Route>
+    )
+  );
+
+  return <RouterProvider router={router} />;
+}
+
 export function GlobalApp(props: GlobalAppProps) {
```
Comment (rart) [inline ui/app/src/components/GlobalApp/GlobalApp.tsx]:
```suggestion
function GlobalAppInternal(props: GlobalAppProps) {
```

File: ui/app/src/pages/Global.tsx [hunk 15]
```diff
@@ -14,15 +14,12 @@
...
-import GlobalApp, { GlobalAppContextProvider } from '../components/GlobalApp';
+import { GlobalAppContextProvider, GlobalAppRouterProvider } from '../components/GlobalApp';
 
 export default function Global(topProps: any) {
   return (
     <GlobalAppContextProvider>
-      <HashRouter>
-        <GlobalApp {...topProps} />
-      </HashRouter>
+      <GlobalAppRouterProvider {...topProps} />
```
Comment (rart) [inline ui/app/src/pages/Global.tsx]:
```suggestion
      <GlobalApp {...topProps} />
```

**Review by rart — CHANGES_REQUESTED** [review]
Do an `upgrade-interactive` and update mui other relevant ones, @jvega190

File: ui/app/src/components/GlobalConfigManagement/GlobalConfigManagement.tsx (hunk 5 above)
Comment (jvega190) [reply ui/app/src/components/GlobalConfigManagement/GlobalConfigManagement.tsx:124]: Modern browsers do not support custom messages in the 'beforeunload' popup. For example see https://chromestatus.com/feature/5349061406228480. The message that shows is:
```
Reload/Leave Site? (depending on action)
Changes you made may not be saved.
```

File: ui/app/src/components/SiteConfigurationManagement/SiteConfigurationManagement.tsx (hunk 12 above)
Comment (jvega190) [reply ui/app/src/components/SiteConfigurationManagement/SiteConfigurationManagement.tsx]: We may need a different way of retrieving this configuration url. Will ping you tomorrow

File: ui/app/src/components/LauncherSection/utils.ts [hunk 16]
```diff
@@ -21,7 +21,7 @@ export const urlMapping = {
   'home.globalMenu.logging-levels': '#/logging',
   'home.globalMenu.log-console': '#/log',
   'home.globalMenu.users': '#/users',
-  'home.globalMenu.sites': '#/sites',
+  'home.globalMenu.sites': '#/projects',
```
Comment (rart) [inline ui/app/src/components/LauncherSection/utils.ts]: I don't remember what exactly is on the global config xml; please check if this new route interferes with it in anyway?

File: ui/app/src/hooks/usePreviewUrlControl.tsx [hunk 17]
```diff
@@ -70,9 +68,9 @@ export function usePreviewUrlControl(history) {
 
         // If site doesn't exist, or its state is not 'READY', alert and navigate to sites page.
         if (!siteExists) {
-          notValidSiteRedirect(formatMessage(messages.siteNotFound), `${authoringBase}#/sites`);
+          notValidSiteRedirect(formatMessage(messages.siteNotFound), `${authoringBase}#/projects`);
```
Comment (rart) [inline ui/app/src/hooks/usePreviewUrlControl.tsx]: Use Enums **everywhere** routes are being used

File: ui/app/src/state/epics/system.ts [hunk 18]
```diff
@@ -475,7 +475,7 @@ const systemEpics: CrafterCMSEpic[] = [
             alert(
               formatMessage({ defaultMessage: "This project has been deleted, you'll be redirected to projects list." })
             );
-            window.location.href = `${authoringBase}#/sites`;
+            window.location.href = `${authoringBase}#/projects`;
```
Comment (rart) [inline ui/app/src/state/epics/system.ts]: enum

**Review by rart — CHANGES_REQUESTED** [review]
Check the Dependabot PRs for newer versions of these updates

File: ui/app/src/components/LauncherSection/utils.ts (hunk 16 above)
Comment (jvega190) [reply ui/app/src/components/LauncherSection/utils.ts]: The global menu config has a list of ids (and some other stuff) matching the keys of the urlMapping const here. So there won't be issues with this as long as we keep the id (or update it everywhere)

```
<globalMenu>
    <items>
        <item>
            <id>home.globalMenu.sites</id>
            <label lang="en">Sites</label>
            <icon>fa-sitemap</icon>
            <permission>*</permission>
        </item>
        ...
```

**Review by rart — APPROVED** [review]

**Pull request merged**